alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+
+    Parameters
+    ----------
+    explainer
+        Fitted explainer.
+    num_prototypes
+        Number of requested prototypes.
+    eps
+        Epsilon radius to be set and used for the computation of prototypes.
+     preprocess_fn
+        Preprocessing function to be applied to the data instance before fitting the 1-KNN classifier
+
+    Returns
+    -------
+    Fitted KNN-classifier.
+    """
+    # update explainer eps and get explanation
+    explainer.eps = eps
+    explanation = explainer.explain(num_prototypes=num_prototypes)
+
+    # train 1-knn classifier
+    proto, proto_labels = explanation.data['prototypes'], explanation.data['prototypes_labels']
+    if len(proto) == 0:
+        return None
+
+    knn = KNeighborsClassifier(n_neighbors=1)
+    return knn.fit(X=proto, y=proto_labels)
+
+
+def cv_protoselect_euclidean(refset: Tuple[np.ndarray, np.ndarray],
+                             protoset: Tuple[np.ndarray, ],
+                             valset: Optional[Tuple[np.ndarray, np.ndarray]] = None,
+                             num_prototypes: int = 1,
+                             eps_range: Optional[np.ndarray] = None,
+                             quantiles: Optional[Tuple[float, float]] = None,
+                             grid_size: int = 25,
+                             n_splits: int = 2,
+                             batch_size: int = int(1e10),
+                             preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+                             **kwargs) -> dict:
+    """
+    Cross-validation parameter selection for ProtoSelect with Euclidean distance. The method computes
+    the best epsilon radius.
+
+    Parameters
+    ----------
+    refset
+        Tuple, `(X_ref, X_ref_labels)`, consisting of the reference data instances with the corresponding reference
+        labels.
+    protoset
+        Tuple, `(X_proto, )`, consisting of the prototypes selection set. Note that the argument is passed as a tuple
+        with a single element for consistency reasons.
+    valset
+        Optional tuple `(X_val, X_val_labels)` consisting of validation data instances with the corresponding
+        validation labels. 1-KNN classifier is evaluated on the validation dataset to obtain the best epsilon radius.
+        In case ``valset=None``, then `n-splits` cross-validation is performed on the `refset`.
+    num_prototypes
+        The number of prototypes to be selected.
+    eps_range
+        Optional ranges of values to select the epsilon radius from. If not specified, the search range is
+        automatically proposed based on the inter-distances between `X_ref` and `X_proto`. The distances are filtered
+        by considering only values in between the `quantiles` values. The minimum and maximum distance values are
+        used to define the range of values to search the epsilon radius. The interval is discretized in `grid_size`
+        equal-distant bins.
```

Comment:
Not sure exactly what `eps_range` should look like. I assume it's not a range of 2 values otherwise it would be of type `Tuple[float, float]`. I see later in code it's actually a `linspace` so it is a set of discrete epsilon values, perhaps need to rephrase the description here?